compiler/rustc_parse/src/parser/item.rs

```diff
@@ -466,22 +487,11 @@ impl<'a> Parser<'a> {
         &mut self,
         attrs: &mut Vec<Attribute>,
         defaultness: Defaultness,
+        // these are passed as params to allow better recovery on `const impl`
+        unsafety: Unsafe,
+        mut generics: Generics,
+        constness: Const,
```

Comment:
I don't think this function needs to be changed?
If you encounter `const impl`, then we can save `const` and its span, then continue parsing using `parse_item_impl`, then patch the result with the saved `const`.
(This logic can be placed to `compiler\rustc_parse\src\parser\diagnostics.rs` to avoid conflating recovery with regular parsing logic.)

Reply:
I did this in order to be able to get the span of the generics as well, in order to provide the diagnostic of `const impl<T>` -> `impl<T> const`. Are you saying to just mutate the output of `parse_item_impl` and get the span from there? I think it'd be preferable to emit the diagnostic prior to parsing the impl if possible, cause then `const impl Foo for >malformed type<` would just error on the type, rather than saying anything about the earlier error that was `const impl`.

Reply:
I meant something like https://github.com/petrochenkov/rust/commit/3d0caf31817f54f79e00518ac9c1387ff5f09cfe (except moved to `diagnostics.rs`), span "before trait" is as good as span "after generics".

>`const impl Foo for >malformed type<` would just error on the syntax error found in the parsing of impl, rather than saying anything about the earlier `const impl` syntax error.

That's totally ok, if you have multiple errors you are not obliged to report all of them.
(Also if `>malformed type<` is recoverable, then both errors will indeed be reported.)